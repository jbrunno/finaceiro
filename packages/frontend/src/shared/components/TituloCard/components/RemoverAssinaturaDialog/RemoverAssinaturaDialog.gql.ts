import { gql } from '@apollo/client';

export const EDITAR_ASSINATURA_DIALOG_MUTATION = gql`
  mutation EditarAssinaturaTitulo($input: EditarAssinaturaTituloInput!) {
    editarAssinaturaTitulo(input: $input) {
      id
    }
  }
`;

export type RemoverAssinaturaDialogData = {
  id: string;
};

export type RemoverAssinaturaDialogInput = {
  id: string;
  usuarioSupervisor: string;
  senhaSupervisor: string;
  motivo: string;
};
