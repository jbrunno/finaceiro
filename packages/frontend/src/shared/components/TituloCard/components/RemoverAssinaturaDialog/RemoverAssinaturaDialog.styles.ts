import { Grid } from '@frontend/components';
import { styled } from '@frontend/styles';

export const GridStyled = styled(Grid)`
  width: ${({ theme }) => theme.fn.size(380)};
`;

export const GridTitle = styled(Grid)`
  padding: ${({ theme }) => theme.fn.spacing(24, 64)};
`;

export const GridForm = styled(Grid)`
  padding: ${({ theme }) => theme.fn.spacing(16, 40, 32)};
`;

export const GridSubmit = styled(Grid)`
  padding: ${({ theme }) => theme.fn.spacing(16, 40, 48)};
`;
