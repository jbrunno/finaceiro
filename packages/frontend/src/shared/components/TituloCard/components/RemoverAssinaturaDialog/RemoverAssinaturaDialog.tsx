import {
  ButtonLoading,
  Dialog,
  Grid,
  PasswordField,
  TextField,
  Typography,
} from '@frontend/components';
import {
  GridForm,
  GridStyled,
  GridSubmit,
  GridTitle,
} from './RemoverAssinaturaDialog.styles';
import {
  FormFields,
  UseRemoverAssinaturaDialogProps,
  useRemoverAssinaturaDialog,
} from './hooks/useRemoverAssinaturaDialog';

type RemoverAssinaturaDialogProps = UseRemoverAssinaturaDialogProps & {
  open: boolean;
};

export function RemoverAssinaturaDialog({
  onClose,
  open,
  tituloId,
  refetch,
}: RemoverAssinaturaDialogProps) {
  const {
    formik,
    handleClose,
    isDisableSubmit,
    isMotivoError,
    isSenhaSupervisorError,
    isUsuarioSupervisorError,
  } = useRemoverAssinaturaDialog({
    onClose,
    tituloId,
    refetch,
  });

  return (
    <Dialog open={open} onClose={handleClose}>
      <GridStyled container>
        <GridTitle item xs={12} alignItems="center" justifyContent="center">
          <Typography variant="h3" textAlign="center">
            Remover status de assinatura do contrato?
          </Typography>
        </GridTitle>
        <form onSubmit={formik.handleSubmit}>
          <GridForm
            item
            xs={12}
            alignItems="center"
            justifyContent="center"
            gap={3}
            container
          >
            <Grid item xs={12}>
              <TextField
                label="Motivo da remoção"
                size="small"
                fullWidth
                value={formik.values.motivo}
                onChange={formik.handleChange(FormFields.motivo)}
                onBlur={formik.handleBlur(FormFields.motivo)}
                error={isMotivoError}
                errorText={formik.errors.motivo}
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                label="Usuário do supervisor"
                size="small"
                fullWidth
                value={formik.values.usuarioSupervisor}
                onChange={formik.handleChange(FormFields.usuarioSupervisor)}
                onBlur={formik.handleBlur(FormFields.usuarioSupervisor)}
                error={isUsuarioSupervisorError}
                errorText={formik.errors.usuarioSupervisor}
              />
            </Grid>
            <Grid item xs={12}>
              <PasswordField
                label="Senha do supervisor"
                size="small"
                fullWidth
                value={formik.values.senhaSupervisor}
                onChange={formik.handleChange(FormFields.senhaSupervisor)}
                onBlur={formik.handleBlur(FormFields.senhaSupervisor)}
                error={isSenhaSupervisorError}
                errorText={formik.errors.senhaSupervisor}
              />
            </Grid>
          </GridForm>
          <GridSubmit item xs={12} alignItems="center" justifyContent="center">
            <ButtonLoading
              type="submit"
              variant="contained"
              color="primary"
              disabled={isDisableSubmit}
              isLoading={false}
              fullWidth
            >
              Remover
            </ButtonLoading>
          </GridSubmit>
        </form>
      </GridStyled>
    </Dialog>
  );
}
