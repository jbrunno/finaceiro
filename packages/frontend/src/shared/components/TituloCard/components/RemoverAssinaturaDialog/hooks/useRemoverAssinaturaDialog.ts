import { useMutation } from '@apollo/client';
import { useFormik } from 'formik';
import * as yup from 'yup';
import { toast } from 'react-toastify';
import { ERRORS_API } from '@/shared/constants';
import {
  EDITAR_ASSINATURA_DIALOG_MUTATION,
  RemoverAssinaturaDialogData,
  RemoverAssinaturaDialogInput,
} from '../RemoverAssinaturaDialog.gql';

export enum FormFields {
  motivo = 'motivo',
  usuarioSupervisor = 'usuarioSupervisor',
  senhaSupervisor = 'senhaSupervisor',
}

const initialValues = {
  [FormFields.motivo]: '',
  [FormFields.usuarioSupervisor]: '',
  [FormFields.senhaSupervisor]: '',
};

const Schema = {
  [FormFields.motivo]: yup.string().required('Campo obrigatório'),
  [FormFields.usuarioSupervisor]: yup.string().required('Campo obrigatório'),
  [FormFields.senhaSupervisor]: yup.string().required('Campo obrigatório'),
};

export type UseRemoverAssinaturaDialogProps = {
  onClose: () => void;
  refetch: (target: 'titulos') => void;
  tituloId: string;
};

export const useRemoverAssinaturaDialog = ({
  onClose,
  tituloId,
  refetch,
}: UseRemoverAssinaturaDialogProps) => {
  const [mutation] = useMutation<
    RemoverAssinaturaDialogData,
    { input: RemoverAssinaturaDialogInput }
  >(EDITAR_ASSINATURA_DIALOG_MUTATION);

  const formik = useFormik({
    initialValues,
    validationSchema: yup.object().shape(Schema),
    onSubmit: async (values) => {
      await mutation({
        variables: {
          input: { ...values, id: tituloId },
        },
      })
        .then(() => {
          handleClose();
          refetch('titulos');
          toast.success('Assinatura removida com sucesso.');
        })
        .catch((error) => {
          const {
            extensions: { statusCode },
          } = error.graphQLErrors[0];

          switch (statusCode) {
            case ERRORS_API.ERROR_401:
              return formik.setErrors({
                usuarioSupervisor: ' ',
                senhaSupervisor: 'Usuário ou senha incorretos.',
              });
            case ERRORS_API.ERROR_403:
              return toast.error('Login do supervisor não autorizado.');
            default:
              return toast.error(
                'Erro ao remover assinatura. Tente novamente.',
              );
          }
        });
    },
  });

  const handleClose = () => {
    formik.resetForm();
    onClose();
  };

  const isDisableSubmit = [
    !formik.values.motivo,
    !formik.values.usuarioSupervisor,
    !formik.values.senhaSupervisor,
  ].includes(true);

  const isMotivoError = !!(formik.touched.motivo && formik.errors.motivo);
  const isUsuarioSupervisorError = !!(
    formik.touched.usuarioSupervisor && formik.errors.usuarioSupervisor
  );
  const isSenhaSupervisorError = !!(
    formik.touched.senhaSupervisor && formik.errors.senhaSupervisor
  );

  return {
    formik,
    handleClose,
    isDisableSubmit,
    isMotivoError,
    isUsuarioSupervisorError,
    isSenhaSupervisorError,
  };
};
