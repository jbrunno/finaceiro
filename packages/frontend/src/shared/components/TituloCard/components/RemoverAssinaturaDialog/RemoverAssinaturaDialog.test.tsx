import { render } from '@test-utils/component-testing';
import { RemoverAssinaturaDialog } from './RemoverAssinaturaDialog';
import { tituloId } from './hooks/useRemoverAssinaturaDialog.mock';

test('should render RemoverAssinaturaDialog', () => {
  const { container } = render(
    <RemoverAssinaturaDialog
      open
      onClose={jest.fn}
      refetch={jest.fn}
      tituloId={tituloId}
    />,
  );
  expect(container).toMatchSnapshot();
});
