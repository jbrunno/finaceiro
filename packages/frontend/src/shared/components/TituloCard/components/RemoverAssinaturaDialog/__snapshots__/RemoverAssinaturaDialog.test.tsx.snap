// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RemoverAssinaturaDialog 1`] = `
<div
  aria-hidden="true"
/>
`;
